### PR TITLE
test: add tests for the `schedule` host fn

### PR DESCRIPTION
**Pull request by ddd-mtl**
https://github.com/holochain/holochain/issues/5139

Added simple tests to check if `schedule` host fn is behaving as expected.

3 tests are ignored as they currently don't pass:
 - When a persisted function returns `None` it should be unscheduled.
 - When a persisted function fails for whatever reason it should be unscheduled.
 - When a persisted function provides an invalid crontab, it should be unscheduled.

## Summary by CodeRabbit

* **Documentation**
  * Clarified scheduling behavior (ephemeral vs. persisted), trigger inputs, error handling, invalid crontab handling, and improved readability; removed outdated guidance; updated changelog to note new scheduling tests.
* **Tests**
  * Added a comprehensive scheduling test suite covering ephemeral and persisted flows, error and edge cases, helpers to observe scheduling state and signals, and conditional inclusion behind a feature flag.

### CODE REVIEW

File: crates/hdk/src/time.rs [hunk 1]
```diff
@@ -71,28 +71,31 @@ pub fn sys_time() -> ExternResult<Timestamp> {
...
-/// whatever was returned by the previous invocation that triggered the current
-/// invocation.
+/// their schedule by returning a different crontab, an ephemeral trigger or `None`
+/// for no further triggers. The initial input schedule of a scheduled function will
+/// be `None`. The following inputs will be whatever was returned by the previous invocation
+/// that triggered the current invocation.
+/// If an invalid crontab is returned, the function will keep its last known valid
+/// schedule. If the last valid schedule is `None`, then it will be unscheduled.
+/// If for any reason the function call fails, it will still be scheduled with the
+/// last known valid schedule.
```
Comment (ddd-mtl) [inline crates/hdk/src/time.rs]: Made doc a bit easier to read and made explicit the behavior when crontab is invalid or zome function call failed.

File: crates/holochain/tests/tests/schedule.rs [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        "integrity",
+        "coordinator",
+        vec![EntryDef::default_from_id("1")],
+        0,
+    )
+    .function::<_, Option<Schedule>, Option<Schedule>>(
+        "coordinator",
+        "ephemeral",
+        move |api, input| {
+            let _ = api.emit_signal(AppSignal::new(ExternIO::encode(input.clone()).unwrap()));
```
Comment (ddd-mtl) [inline crates/holochain/tests/tests/schedule.rs]: function emits a signal of its input as of way of signaling it has been called.

File: crates/holochain/tests/tests/schedule.rs [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+            let ms: u64 = match input {
+                Some(Schedule::Ephemeral(duration)) => duration.as_millis() as u64,
+                None => 3,
+                _ => panic!("Expected Ephemeral Schedule"),
+            };
+            match ms {
+                0 => Ok(None),
+                _ => Ok(Some(Schedule::Ephemeral(std::time::Duration::from_millis(
+                    ms - 1,
```
Comment (ddd-mtl) [inline crates/holochain/tests/tests/schedule.rs]: Decrease duration by 1ms after each call and stop at zero.

File: crates/holochain/tests/tests/schedule.rs [hunk 4]
```diff
@@ -0,0 +1,286 @@
...
+    let app = conductor.setup_app("app", &[dna.0]).await.unwrap();
+    let cell = app.into_cells()[0].clone();
+    let mut app_signal = conductor.subscribe_to_app_signals("app".into());
+
+    conductor
+        .call::<(), ()>(&cell.zome("coordinator"), "start", ())
+        .await;
+
+    // Scheduled function is first called with None input
+    assert_eq!(None, wait_for_signal(&mut app_signal).await.unwrap());
```
Comment (veeso) [inline crates/holochain/tests/tests/schedule.rs:104]:
```suggestion
    assert!(wait_for_signal(&mut app_signal).await.is_none());
```

File: crates/holochain/tests/tests/schedule.rs [hunk 5]
```diff
@@ -0,0 +1,286 @@
...
+    let cell = app.into_cells()[0].clone();
+    let mut app_signal = conductor.subscribe_to_app_signals("app".into());
+
+    // Start test: schedule function
+    conductor
+        .call::<(), ()>(&cell.zome("coordinator"), "start", ())
+        .await;
+
+    // Scheduled function is first called with None input.
+    assert_eq!(None, wait_for_signal(&mut app_signal).await.unwrap());
```
Comment (veeso) [inline crates/holochain/tests/tests/schedule.rs:153]:
```suggestion
    assert!(wait_for_signal(&mut app_signal).await.is_none());
```

File: crates/holochain/tests/tests/schedule.rs [hunk 6]
```diff
@@ -0,0 +1,286 @@
...
```
Comment (veeso) [inline crates/holochain/tests/tests/schedule.rs:240]:
```suggestion
    assert!(wait_for_signal(&mut app_signal).await.is_none());
```

File: crates/holochain/tests/tests/schedule.rs [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+        HostFnCaller::create_for_zome(cell.cell_id(), &conductor.raw_handle(), &dna.0, 0).await;
+
+    // Start test: schedule function
+    conductor
+        .call::<(), ()>(&cell.zome("coordinator"), "start", ())
+        .await;
+    // Should be scheduled
+    assert!(is_scheduled(&host_fn_caller, &cell).await);
+    // Scheduled function is first called with None input.
+    assert_eq!(None, wait_for_signal(&mut app_signal).await.unwrap());
```
Comment (veeso) [inline crates/holochain/tests/tests/schedule.rs:281]:
```suggestion
    assert!(wait_for_signal(&mut app_signal).await.is_none());
```

File: crates/holochain/tests/tests/mod.rs [hunk 8]
```diff
@@ -21,6 +21,8 @@ mod new_lair;
 mod paths;
 mod publish;
 mod regression;
+#[cfg(feature = "unstable-functions")]
+mod schedule;
```
Comment (ThetaSinner) [inline crates/holochain/tests/tests/mod.rs:25]: This is a good place to add tests like these. There are a lot of sweettests mixed into the codebase where ideally only unit tests would be and it makes it slow to run just the unit tests on Holochain. I'd really like to move all our integration tests out here.

File: crates/holochain/tests/tests/schedule.rs [hunk 9]
```diff
@@ -0,0 +1,286 @@
...
+#[tokio::test(flavor = "multi_thread")]
+async fn schedule_ephemeral() {
+    holochain_trace::test_run();
+
+    // Start with a duration of 3ms and decrease by 1ms each time it's called.
+    let zome = create_schedule_zome(|api, input| {
+        let _ = api.emit_signal(AppSignal::new(ExternIO::encode(input.clone()).unwrap()));
+        let ms: u64 = match input {
+            Some(Schedule::Ephemeral(duration)) => duration.as_millis() as u64,
+            None => 3,
```
Comment (ThetaSinner) [inline crates/holochain/tests/tests/schedule.rs]:
```suggestion
            None => 3,
            Some(Schedule::Ephemeral(duration)) => duration.as_millis() as u64,
```

Possibly easier to follow? The `None` is expected first, then we take the input value on subsequent schedules calls.

File: crates/holochain/tests/tests/schedule.rs (hunk 4 above)
Comment (ThetaSinner) [reply crates/holochain/tests/tests/schedule.rs:104]: That is more compact and I was wondering about it above but actually it's nice to see what's inside the `Some` if this fails

**Review by ThetaSinner — COMMENTED** [review]
The testing approach looks sound. What I'm unclear about is why all the tests pass. Are the tests being written to pass what is currently happening or is there missing test coverage for the problem cases?

Increasing test coverage was the point of this so if the problem cases covered then that's okay. They could be added as ignored but they could also be added later.

If the tests are testing the current state of things and the assertions aren't matching the desired end state then I'd prefer assertions that fail and to just mark the tests as `#[ignore]` for now until we fix the problems.

**Comment by ThetaSinner** [comment]
> When a persisted function returns None it should be unscheduled.

Yes

> An ephemeral duration of 0 is deemed acceptable and the function stays scheduled.

I think that is a little odd but probably acceptable if documented.

> A persisted function that fails for whatever reason will still stay scheduled.

That's a tricky one because it could be that the error is systemic and will just keep happening so we'd be better off unscheduling. That's a more complicated change if we need to modify the required return values and behavior of scheduled functions. I think this would be worth some investigation for what would need to change and then discussion?

> When a persisted function changes to provide an invalid crontab, it will still stay scheduled.

Again, hard to tell whether the problem is systemic or temporary. I think demanding valid crontab being returned is reasonable and the schedule should be cancelled.

File: crates/holochain/tests/tests/mod.rs (hunk 8 above)
Comment (ddd-mtl) [reply crates/holochain/tests/tests/mod.rs:25]: Does that work for you if I move the pre-existing schedule tests in that file in a subsequent PR?

**Comment by ddd-mtl** [comment]
@ThetaSinner Changed tests to match expectations and set to `ignore` when failing. Changed doc accordingly.

Added ephemeral_error test and persisted_expired test to cover more cases.

**Review by matthme — COMMENTED** [review]
Changelog in `crates/holochain/CHANGELOG.md` is missing. It's "only" adding tests but I think it's still worth mentioning in the changelog.

**Comment by ddd-mtl** [comment]
@matthme changelog item added

File: crates/holochain/tests/tests/schedule.rs [hunk 10]
```diff
@@ -0,0 +1,380 @@
...
+    // On bad crontab scheduled function should unschedule
+    assert!(wait_for_signal(&mut app_signal).await.is_err());
+    assert!(!is_scheduled(&host_fn_caller, &cell).await);
+}
+
+/// Test schedule persisted function which gives an error
+/// Assuming a schedular interval of 100ms
+#[tokio::test(flavor = "multi_thread")]
+#[ignore = "Test not passing ; awaiting bug fix"]
+async fn schedule_persisted_error() {
```
Comment (matthme) [inline crates/holochain/tests/tests/schedule.rs]:
```suggestion
async fn persisted_function_that_errors_gets_unscheduled() {
```

Comment (matthme) [reply crates/holochain/tests/tests/schedule.rs]: I think the function name or at least the documentation of the test function should describe the expected behavior.

File: crates/holochain/tests/tests/schedule.rs [hunk 11]
```diff
@@ -0,0 +1,380 @@
...
+async fn is_scheduled(host_fn_caller: &HostFnCaller, cell: &SweetCell) -> bool {
+    let pubkey = cell.cell_id().agent_pubkey().clone();
+    host_fn_caller
+        .authored_db
+        .read_async({
+            move |txn| {
+                Result::<bool, DatabaseError>::Ok(
+                    fn_is_scheduled(
+                        txn,
+                        ScheduledFn::new("coordinator".into(), "scheduled".into()),
```
Comment (matthme) [inline crates/holochain/tests/tests/schedule.rs]: I think it would be a bit more resilient to take out the zome and function names here as constants and reuse the constants both here in `is_scheduled` and in `create_schedule_zome`. Or potentially even more resilient, to take the zome and function name as input arguments to `is_sheduled`. Otherwise there's a possibility that if further tests may get added in the future that someone uses the `is_sheduled` function to unknowingly check against the wrong zome function.

File: crates/holochain/tests/tests/schedule.rs [hunk 12]
```diff
@@ -0,0 +1,380 @@
...
+    // Should be unscheduled
+    assert!(wait_for_signal(&mut app_signal).await.is_err());
+    assert!(!is_scheduled(&host_fn_caller, &cell).await);
+}
+
+/// Test persisted schedule with invalid crontab output
+/// Assuming a schedular interval of 100ms
+#[tokio::test(flavor = "multi_thread")]
+#[ignore = "Test not passing ; awaiting bug fix"]
+async fn schedule_persisted_bad_crontab() {
```
Comment (matthme) [inline crates/holochain/tests/tests/schedule.rs]:
```suggestion
async fn persisted_function_with_bad_crontab_gets_unscheduled() {
```

File: crates/holochain/tests/tests/schedule.rs [hunk 13]
```diff
@@ -0,0 +1,380 @@
...
+use holochain_types::inline_zome::InlineZomeSet;
+use holochain_types::signal::Signal;
+use holochain_zome_types::prelude::Schedule::{Ephemeral, Persisted};
+use holochain_zome_types::prelude::{Schedule, ScheduledFn};
+use holochain_zome_types::signal::AppSignal;
+use tokio::sync::broadcast;
+use tokio::time::error::Elapsed;
+
+/// Test schedule ephemeral fn
+/// Assuming a schedular interval of 100ms
```
Comment (matthme) [inline crates/holochain/tests/tests/schedule.rs]:
```suggestion
/// Assuming a scheduler interval of 100ms
```

File: crates/holochain/tests/tests/schedule.rs [hunk 14]
```diff
@@ -0,0 +1,380 @@
...
+    assert_eq!(
+        Some(Schedule::Ephemeral(std::time::Duration::from_millis(0))),
+        wait_for_signal(&mut app_signal).await.unwrap()
+    );
+    assert!(wait_for_signal(&mut app_signal).await.is_err());
+    assert!(!is_scheduled(&host_fn_caller, &cell).await);
+}
+
+/// Test schedule ephemeral function which gives an error
+/// Assuming a schedular interval of 100ms
```
Comment (matthme) [inline crates/holochain/tests/tests/schedule.rs]:
```suggestion
/// Assuming a scheduler interval of 100ms
```

File: crates/holochain/tests/tests/schedule.rs [hunk 15]
```diff
@@ -0,0 +1,380 @@
...
+    assert_eq!(None, wait_for_signal(&mut app_signal).await.unwrap());
+    // Scheduled function is called with input from previous output.
+    assert!(wait_for_signal(&mut app_signal).await.unwrap().is_some());
+    // Should be unscheduled
+    assert!(wait_for_signal(&mut app_signal).await.is_err());
+    assert!(!is_scheduled(&host_fn_caller, &cell).await);
+}
+
+/// Test schedule persisted function with changing crontab
+/// Assuming a schedular interval of 100ms
```
Comment (matthme) [inline crates/holochain/tests/tests/schedule.rs]:
```suggestion
/// Assuming a scheduler interval of 100ms
```

File: crates/holochain/tests/tests/mod.rs (hunk 8 above)
Comment (ThetaSinner) [reply crates/holochain/tests/tests/mod.rs:25]: Yes that'd be great. Any time we can move a few tests to tidy up is helpful.

File: crates/holochain/tests/tests/schedule.rs [hunk 16]
```diff
@@ -0,0 +1,380 @@
...
+#[tokio::test(flavor = "multi_thread")]
+#[ignore = "Test not passing ; awaiting bug fix"]
+async fn schedule_persisted_ok() {
```
Comment (ThetaSinner) [inline crates/holochain/tests/tests/schedule.rs]: Could be clearer from the title what the test is doing, something like `schedule_persisted_then_unschedule`. Keeping something scheduled forever is also an "ok" path.

File: crates/holochain/tests/tests/schedule.rs (hunk 12 above)
Comment (ddd-mtl) [reply crates/holochain/tests/tests/schedule.rs]: Renamed to `schedule_persisted_fn_with_bad_crontab()` and added expected behavior in comment

File: crates/holochain/tests/tests/schedule.rs (hunk 10 above)
Comment (ddd-mtl) [reply crates/holochain/tests/tests/schedule.rs]: Renamed to `schedule_persisted_fn_that_errors()` and added expected behavior to doc comment

File: crates/holochain/tests/tests/schedule.rs (hunk 11 above)
Comment (ddd-mtl) [reply crates/holochain/tests/tests/schedule.rs]: Changed to consts. Input arguments is more generic but actually less resilient since bad inputs are made possible.

File: crates/holochain/tests/tests/schedule.rs (hunk 16 above)
Comment (ddd-mtl) [reply crates/holochain/tests/tests/schedule.rs]: Renamed to `schedule_persisted_fn_then_unschedule()` and edited doc comment
